src/pybind/mgr/telemetry/module.py

```diff
@@ -1033,52 +1482,171 @@ def do_send(self,
-        try:
-            # Formatting ranges and values in osd_perf_histograms
-            modes_to_be_formatted = ['osd_perf_histograms_aggregated', 'osd_perf_histograms_separated']
-            for mode in modes_to_be_formatted:
-                for config in report[mode]:
-                    for histogram in config:
-                        # Adjust ranges by converting lists into strings
-                        for axis in config[histogram]['axes']:
-                            for i in range(0, len(axis['ranges'])):
-                                axis['ranges'][i] = str(axis['ranges'][i])
-                        # Adjust values by converting lists into strings
-                        if mode == 'osd_perf_histograms_aggregated':
-                            for i in range(0, len(config[histogram]['values'])):
-                                config[histogram]['values'][i] = str(config[histogram]['values'][i])
-                        else: # if mode == 'osd_perf_histograms_separated'
-                            for osd in config[histogram]['osds']:
-                                for i in range(0, len(osd['values'])):
-                                    osd['values'][i] = str(osd['values'][i])
-        except KeyError:
-            # If the perf channel is not enabled, there should be a KeyError since
-            # 'osd_perf_histograms' would not be present in the report. In that case,
-            # the show function should pass as usual without trying to format the
-            # histograms.
-            pass
+        if self.channel_device:
+            report += '''\nDevice report is generated separately. To see it run 'ceph telemetry show-device'.'''
 
+        return 0, report, ''
+
+    @CLIReadCommand('telemetry preview')
+    def preview(self, channels: Optional[List[str]] = None) -> Tuple[int, str, str]:
+        '''
+        Preview a sample report of the most recent collections available (except for 'device')
+        '''
+        report = {}
+
+        # We use a lock to prevent a scenario where the user wishes to preview
+        # the report, and at the same time the module hits the interval of
+        # sending a report with the opted-in collection, which has less data
+        # than in the preview report.
+        col_delta = self.collection_delta()
+        with self.get_report_lock:
+            if col_delta is not None and len(col_delta) == 0:
+                # user is already opted-in to the most recent collection
+                msg = 'Telemetry is up to date, see report with `ceph telemetry show`.'
+                return 0, msg, ''
+            else:
+                # there are collections the user is not opted-in to
+                next_collection = []
+
+                for c in MODULE_COLLECTION:
+                    for k, v in c.items():
+                        if k == 'name':
+                            next_collection.append(v.name)
```

Comment:
This piece of code is duplicated a lot. Would it be worth it to write a smaller function to encapsulate this? Something like `update_module_collection()`.
If you don't think it's needed though, ignore this comment.

Reply:
as above, I'd suggest using dict lookup instead of a loop, then there's little duplication